### PR TITLE
Forgot password

### PR DESCRIPTION
**Pull request by jonathanmcdaniel**
This PR allows a user to reset their password through a series of steps.

**Significant Changes**
1. New email template
2. New attribute being added to a user item in the DB (only for the duration between forgot password and reset password)
3. Removal of test auth pages from the very beginning of the project
4. Some cleanup on sign in and register

**How to Test**
This is easiest to test using two different browsers at the same time.
1. Open Browser1 (B1) and Browser2 (B2)
2. Create a new account using B1
3. Activate account
4. Sign in with credentials in B1
5. Use B2 to click forgot password on the sign in page
6. Enter email used during registration
7. Open email and copy link to reset password
8. Open this link in B2
9. Reset the password for that user
10. Try to log in with the new password using B2 (should work)
11. Try to click around in B1 and you should be kicked out and sent back to login (since that auth token is no longer valid).

**Other Things to Test**
1. Try changing the password of a different user.
2. Try using the forgot password link more than once. The link itself will work but the reset password functionality will not.

**Notes**
😸

### CODE REVIEW

**Comment by smc395** [comment]
Can you change the landing page message after a user submits an email to "A link has been sent to the email address provided. If you did not receive this email, please check your spam or try again."

File: Backend/mentii/user_ctrl.py [hunk 1]
```diff
@@ -10,6 +10,65 @@
...
+  url = host + '/reset-password/{0}'.format(resetPasswordId)
+  message = render_template('forgotPasswordEmail.html', url=url)
+  #Build Message
+  msg = Message('Mentii: Reset Password', recipients=[email], extra_headers={'Content-Transfer-Encoding': 'quoted-printable'}, html=message)
+  #Send Email
+  mailer.send(msg)
+  return response
+
+def addResetPasswordIdToUser(email, resetPasswordId, dbInstance):
+  table = dbUtils.getTable('users', dbInstance)
```
Comment (smc395) [inline Backend/mentii/user_ctrl.py]: Missing check if table is None

File: Backend/mentii/user_ctrl.py [hunk 2]
```diff
@@ -10,6 +10,65 @@
...
+  jsonData = {
+    'Key': {'email': email},
+    'UpdateExpression': 'SET resetPasswordId = :a',
+    'ExpressionAttributeValues': { ':a': resetPasswordId },
+    'ReturnValues' : 'UPDATED_NEW'
+  }
+  res = dbUtils.updateItem(jsonData, table)
```
Comment (smc395) [inline Backend/mentii/user_ctrl.py]: `res =` is unnecessary

File: Backend/mentii/user_ctrl.py [hunk 3]
```diff
@@ -10,6 +10,65 @@
...
+
+def resetUserPassword(jsonData, dbInstance):
+  response = ControllerResponse()
+  email = jsonData.get('email', None)
+  password = jsonData.get('password', None)
+  resetPasswordId = jsonData.get('id', None)
+  if email is not None or password is not None or resetPasswordId is not None:
```
Comment (smc395) [inline Backend/mentii/user_ctrl.py]: Should these be `and` clauses?

File: Backend/tests/test_user_ctrl.py [hunk 4]
```diff
@@ -327,6 +327,20 @@ def test_joinClass(self):
     #confirm none joined
     self.assertFalse(joined)
 
+  def test_resetPassword(self):
```
Comment (smc395) [inline Backend/tests/test_user_ctrl.py]: Technically all methods in `user_ctrl` that are written should be unit tested. Does this hit all branches for other methods?

File: Backend/mentii/user_ctrl.py [hunk 5]
```diff
@@ -10,6 +10,65 @@
 import hashlib
 from flask import g
 
+def sendForgotPasswordEmail(httpOrigin, jsonData, mailer, dbInstance):
+  email = jsonData.get('email', None)
+  resetPasswordId = str(uuid.uuid4())
+  addResetPasswordIdToUser(email, resetPasswordId, dbInstance)
```
Comment (RyanYoung25) [inline Backend/mentii/user_ctrl.py]: This function call could fail, (currently uncaught in the function).

File: Backend/mentii/user_ctrl.py (hunk 1 above)
Comment (RyanYoung25) [inline Backend/mentii/user_ctrl.py]: Please add a Boolean return, true for success and false for fail at the end of the function.

File: Backend/app.py [hunk 6]
```diff
@@ -301,6 +301,27 @@ def updateClassDetails():
...
   return ResponseCreation.createResponse(res,status)
 
+@app.route('/forgotPassword/', methods=['POST', 'OPTIONS'])
+@handleOptionsRequest
+def forgotPassword():
+  status = 200
+  res = ResponseCreation.ControllerResponse()
+  dynamoDBInstance = getDatabaseClient()
+  httpOrigin = request.environ.get('HTTP_ORIGIN')
+  res = user_ctrl.sendForgotPasswordEmail(httpOrigin, request.json, mail, dynamoDBInstance)
```
Comment (RyanYoung25) [inline Backend/app.py]: This function could fail after making the updates Micah pointed out. Remember to update this to handle if res.hasErrors()

File: Backend/mentii/user_ctrl.py (hunk 3 above)
Comment (RyanYoung25) [inline Backend/mentii/user_ctrl.py]: +1

File: Backend/mentii/user_ctrl.py (hunk 5 above)
Comment (jonathanmcdaniel) [inline Backend/mentii/user_ctrl.py]: I want this function to fail completely silently as to not hint if a user exists or not. This has been reflected by always returning an empty response in app.py.

File: Backend/tests/test_user_ctrl.py (hunk 4 above)
Comment (jonathanmcdaniel) [inline Backend/tests/test_user_ctrl.py]: I wrote more tests to cover more, but the function that actually sends the email cannot be tested for obvious reasons

**Pull request merged**